### PR TITLE
Put off major collections to improve GC times.

### PR DESCRIPTION
**Pull request by chflood**
Put off full GC unless absolutely necessary.  This fixes some egregious behavior as in https://github.com/JuliaLang/julia/issues/40644 but uses more memory.

Before:
 1.635486 seconds (100.00 k allocations: 863.182 MiB, 73.03% gc time)
  1.318420 seconds (100.00 k allocations: 863.182 MiB, 74.84% gc time)
  1.075697 seconds (100.00 k allocations: 863.182 MiB, 65.05% gc time)
  1.474924 seconds (100.00 k allocations: 863.182 MiB, 73.06% gc time)
  1.100034 seconds (100.00 k allocations: 863.182 MiB, 66.54% gc time)
  1.362786 seconds (100.00 k allocations: 863.182 MiB, 79.13% gc time)
  1.417983 seconds (100.00 k allocations: 863.182 MiB, 74.69% gc time)
  1.046719 seconds (100.00 k allocations: 863.182 MiB, 65.62% gc time)
  1.347786 seconds (100.00 k allocations: 863.182 MiB, 78.98% gc time)
  1.004516 seconds (100.00 k allocations: 863.182 MiB, 68.80% gc time)

After:  1.447965 seconds (100.00 k allocations: 863.182 MiB, 63.36% gc time)
  0.544911 seconds (100.00 k allocations: 863.182 MiB, 2.97% gc time)
  0.531376 seconds (100.00 k allocations: 863.182 MiB)
  0.387158 seconds (100.00 k allocations: 863.182 MiB, 6.38% gc time)
  0.441621 seconds (100.00 k allocations: 863.182 MiB, 16.72% gc time)
  0.504395 seconds (100.00 k allocations: 863.182 MiB)
  0.472456 seconds (100.00 k allocations: 863.182 MiB)
  0.660828 seconds (100.00 k allocations: 863.182 MiB, 53.38% gc time)
  0.289179 seconds (100.00 k allocations: 863.182 MiB)
  0.350453 seconds (100.00 k allocations: 863.182 MiB, 17.39% gc time)

### CODE REVIEW

File: src/gc-debug.c [hunk 1]
```diff
@@ -977,6 +977,15 @@ void gc_time_sweep_pause(uint64_t gc_end_t, int64_t actual_allocd,
                    jl_ns2ms(gc_postmark_end - gc_premark_end),
                    sweep_full ? "full" : "quick", -gc_num.allocd / 1024);
 }
+
+void gc_time_summary(int sweep_full, uint64_t start, uint64_t end, uint64_t freed, uint64_t live, uint64_t interval, uint64_t pause){
+  if (sweep_full > 0)
```
Comment (JeffBezanson) [inline src/gc-debug.c:982]: 4 space indent

**Comment by JeffBezanson** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/44215#issuecomment-1042380078):
```
NanosoldierError: error when preparing/pushing to report repo: failed process: Process(setenv(`git push`; dir="/nanosoldier/workdir/NanosoldierReports"), ProcessExited(1)) [1]

```
Unfortunately, the logs could not be uploaded.

**Comment by vtjnash** [comment]
Looks like noise https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/75f4f26_vs_7889b2a/report.md. Note that we have very few gc-benchmarks, other than the test harness itself (which is not measured).

File: src/gc.c [hunk 2]
```diff
@@ -3140,44 +3139,27 @@ static int _jl_gc_collect(jl_ptls_t ptls, jl_gc_collection_t collection)
...
+      if (not_freed_enough) {
+	gc_num.interval = gc_num.interval * 2;
+      }
+
+      if (large_frontier) {
+	sweep_full = 1;
+      }
+
+      if (gc_num.interval > max_collect_interval) {
+	if (not_freed_enough)
```
Comment (JeffBezanson) [inline src/gc.c:3159]: Since we only grow the interval when `not_freed_enough` is true, is this redundant?

Comment (chflood) [reply src/gc.c]: I don't remember why I thought that was necessary, but you are right.

**Pull request merged**